sw/device/tests/dif/dif_uart_test_mocked.cc

```diff
@@ -0,0 +1,734 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cassert>
+
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+
+extern "C" {
+#include "sw/device/lib/dif/dif_uart.h"
+#include "uart_regs.h"  // Generated.
+}  // extern "C"
+
+namespace {
+constexpr uint32_t kUint32Max = 0xffffffff;
+constexpr uint8_t kDifUartInterruptCount = kDifUartInterruptLast + 1;
+
+typedef struct DifUartInterruptPair {
+  dif_uart_interrupt_t name;
+  uint32_t value;
+} DifUartInterruptPairStruct;
+
+// INTR_STATE, INTR_ENABLE, INTR_TEST registers have interrupt bits in the
+// same sequence.
+constexpr std::array<DifUartInterruptPairStruct, kDifUartInterruptCount>
+    kUartIrqRegArray = {{
+        {kDifUartInterruptTxWatermark, 0b00000001},
+        {kDifUartInterruptRxWatermark, 0b00000010},
+        {kDifUartInterruptTxEmpty, 0b00000100},
+        {kDifUartInterruptRxOverflow, 0b00001000},
+        {kDifUartInterruptRxFrameErr, 0b00010000},
+        {kDifUartInterruptRxBreakErr, 0b00100000},
+        {kDifUartInterruptRxTimeout, 0b01000000},
+        {kDifUartInterruptRxParityErr, 0b10000000},
+    }};
+
+class DifUartTest : public testing::Test, public mock_mmio::MmioTest {
+ protected:
+  dif_uart_t dif_uart_;
+
+  // NCO is calculated as NCO = 2^20 * fbaud / fpclk, so the following values
+  // should result in NCO of 1. This is the default configuration, which will
+  // be used unless the values are overriden.
+  dif_uart_config_t dif_uart_config_ = {
+      1,                  /* baudrate */
+      1048576,            /* clk_freq_hz */
+      kDifUartDisable,    /* parity_enable */
+      kDifUartParityEven, /* parity */
+  };
+
+  /**
+   * Initialises the UART state structure without calling the dif_uart_init()
+   * routine.
+   */
+  void DeviceInitMinimal(void) { dif_uart_.base_addr = dev().region(); }
+
+  /**
+   * dif_uart_init() must always result in sequence of these calls before any
+   * other initialisation/configuration register access.
+   */
+  void DeviceResetExpectations(void) {
```

Comment:
If you want to wrap a bunch of EXPECTs, consider a name like `ExpectReset()`.

Reply:
It is still pretty easy to determine where the fault is, as it will tell you which test failed. So I will use expects in functions for now, and maybe I will change my mind later.